### PR TITLE
Make mods able to customize flairs

### PR DESCRIPTION
**Pull request by 13steinj**
As requested on /r/ModSupport:

Allows mods with the flair perm to customize the class of flairs. Also, if the user is a mod, always have a default template to choose from.

This also corrects a small UI bug which can be reproduced as mentioned [here](https://redd.it/3q3z17), as well as another tiny UI issue regarding spacing of the submit button, and the status span.

Edit: Before considering this, please consider #1423, as that fixes a bug, and if that gets merged, a minor change would have to be made here as well.

### CODE REVIEW

File: r2/r2/public/static/css/reddit.less [hunk 1]
```diff
@@ -1031,7 +1031,9 @@ a.banned-user { color: red; }
 }
 
 .flairselector .customizer { display: inline-block; }
-.flairselector .customizer input { display: none; }
+.flairselector .customizer input,
+.flairselector .mod.customizer span { display: none; }
+.flairselector .mod.customizer span { color: black; }
```
Comment (13steinj) [inline r2/r2/public/static/css/reddit.less]: Not really needed; but it makes things look nicer for the user flair selector.

File: r2/r2/lib/pages/pages.py [hunk 2]
```diff
@@ -4029,6 +4031,11 @@ def __init__(self, user, site, link=None):
         else:
             choices = []
 
+        if admin:
+            admintempl = FlairTemplate()
+            admintempl.text = "ModFlair"
+            choices.append(target_wrapper(admintempl))
```
Comment (13steinj) [inline r2/r2/lib/pages/pages.py]: This is essentially, a "FakeFlairTemplate"; and there is no worry about it's id being none; as this only occurs for mods / admins; and those inputs never require an ID.

File: r2/r2/controllers/api.py [hunk 3]
```diff
@@ -3743,6 +3763,8 @@ def POST_flair(self, form, jquery, user, link, text, css_class):
                 jquery.redirect('?name=%s' % user.name)
                 return
 
+            # XXX: more gross UI code
+            # Push some client-side updates back to the browser.
```
Comment (13steinj) [inline r2/r2/controllers/api.py]: Left as a reminder like in POST_selectflair

File: r2/r2/templates/flairselector.html [hunk 4]
```diff
@@ -31,26 +31,48 @@
               li_class += ' texteditable'
           if choice.flair_template_id == thing.matching_template:
               li_class += ' selected'
+          choice.flair_css = choice.flair_css_class if thing.target_type == "link" else choice.flair_css_class[6:]
```
Comment (13steinj) [inline r2/r2/templates/flairselector.html]: User flairs and their css classes are weird; and that's why this is needed.

File: r2/r2/templates/flairselector.html [hunk 5]
```diff
@@ -31,26 +31,48 @@
...
+      <div class="flairselection">
+        <div class="flairremove">
+          (<a href="javascript://void(0)">${_('remove flair')}</a>)
+        </div>
+      </div>
+      <input type="hidden" name="flair_template_id">
+      <div class="customizer">
+        <input type="text" size="16" maxlength="64" name="text">
+      </div>
+      </br>
```
Comment (13steinj) [inline r2/r2/templates/flairselector.html]: Without this </br>; there is a UI issue regarding overflow when the status is displayed.

Comment (wodim) [inline r2/r2/templates/flairselector.html]: You probably meant `<br/>`.

Comment (13steinj) [inline r2/r2/templates/flairselector.html]: Yup. Thanks for the catch. I'll rebase at home.

**Comment by umbrae** [comment]
Can you explain more the intent behind this and some examples of how it works? Screenshots also would be handy if you have them.

(Also just as an aside, we're pretty strict nowadays about not letting in code with TODO or XXX blocks in them. Probably don't fix them just yet so that we can determine if we think this is something we need, but if this is something that looks good we'll probably want to tackle those.)

Thanks for your contributions as usual!

**Comment by 13steinj** [comment]
@umbrae 

The intent behind this was a very recent /r/ModSupport post. As mods, there are times when flairs are needed to be mod only, so that no user can access them. While this is extremely easy for text alone, as a mod with flair perms can edit any flair's text; it is not easy for subreddit design and style.

To make mod only link flairs, as of right now, mods have to set up a comment tagging system using automoderater. It's good, but also is a bit of an inconvenicence. And manual selection of css class is it's own issue.

The base of the pr is this: change the endpoint to used to /post/flair, and add the proper code that would make client side updates there. This way, any mod with the flair perm can edit flairs' css class with no hassle whatsoever.

The added lines by the note [here](https://github.com/reddit/reddit/pull/1450#discussion_r42954010) make a flair template available to mods no matter what, even if there are no flairs, and allows for a blank slate css class wise and a semi blank slate text wise (text is generically "ModFlair").

Of course; this is just the basics. If you'd like, which you probably would since the blank slate flair is kinda done weirdly, I can expand this to actually make mod only flair templates in general, and make an entirely new selection underneath the non-mod's flair selection, which would definitely look nicer though I don't know how / if it should follow the limit on the amount of possible flair templates (which IIRC is 304? I can't find the exact line atm)

Taking screenshots is probably going to be difficult but should be doable. I'll get back when I clean a few small things up elsewhere.

**Comment by umbrae** [comment]
Thanks for the contribution - at this stage I think this isn't a priority for us and we don't have enough time to think through its implications, so I'm going to close this.

I know it's a bummer to write a patch this large and have it closed on you, sorry about that.

**Comment by 13steinj** [comment]
@umbrae Nah it's cool :P

**Pull request closed without merging**